gnu_dataset/mifluz-0.26.0/man/WordType.3

'''
''' Part of the ht://Dig package   <http://www.htdig.org/>
''' Copyright (c) 1999, 2000, 2001 The ht://Dig Group
''' For copyright details, see the file COPYING in your distribution
''' or the GNU General Public License version 2 or later
''' <http://www.gnu.org/copyleft/gpl.html>
''' 
''' 
.TH WordType 3 local
.SH NAME
WordType \-
defines a word in term of allowed characters, length etc.


.SH SYNOPSIS
.nf
.ft CW

Only called thru WordContext::Initialize()
.ft R
.fi

.SH DESCRIPTION

WordType defines an indexed word and operations to validate
a word to be indexed. All words inserted into the
.I mifluz
index
are
.B Normalize
d before insertion. The configuration options
give some control over the definition of a word.


.SH CONFIGURATION
For more information on the configuration attributes and a complete list of attributes, see the mifluz(3) manual page.
.TP
.B  wordlist_locale <locale> (default C)
Set the locale of the program to
.B locale
. See setlocale(3)
for more information.
.TP
.B  wordlist_allow_numbers {true|false} <number> (default false)
A digit is considered a valid character within a word if
this configuration parameter is set to
.I true
otherwise
it is an error to insert a word containing digits.
See the
.B Normalize
method for more information.
.TP
.B  wordlist_mimimun_word_length <number> (default 3)
The minimum length of a word.
See the
.B Normalize
method for more information.
.TP
.B  wordlist_maximum_word_length <number> (default 25)
The maximum length of a word.
See the
.B Normalize
method for more information.
.TP
.B  wordlist_allow_numbers {true|false} <number> (default false)
A digit is considered a valid character within a word if
this configuration parameter is set to
.I true
otherwise
it is an error to insert a word containing digits.
See the
.B Normalize
method for more information.
.TP
.B  wordlist_truncate {true|false} <number> (default true)
If a word is too long according to
the
.I wordlist_maximum_word_length
it is truncated
if this configuration parameter is
.I true
otherwise it
is considered an invalid word.
.TP
.B  wordlist_lowercase {true|false} <number> (default true)
If a word contains upper case letters it is converted to lowercase
if this configuration parameter is true, otherwise it is left
untouched.
.TP
.B  wordlist_valid_punctuation [characters] (default none)
A list of punctuation characters that may appear in a word. 
These characters will be removed from the word before insertion
in the index.
.PP

.SH METHODS
.TP
.B   int Normalize(String &s) const
Normalize a word according to configuration specifications and 
builtin transformations.
.B Every
word inserted in the inverted
index goes thru this function. If
a word is rejected (return value has WORD_NORMALIZE_NOTOK bit set) it will not 
be inserted in the index. If a word is accepted (return value has 
WORD_NORMALIZE_OK bit set) it will be inserted in the index. In
addition to these two bits, informational values are stored that
give information on the processing done on the word.
The bit field values and their meanings are
as follows:
.TP
.B WORD_NORMALIZE_TOOLONG
the word length exceeds the value of 
    the
.I wordlist_maximum_word_length
configuration parameter.
.TP
.B WORD_NORMALIZE_TOOSHORT
the word length is smaller than the value of 
    the
.I wordlist_minimum_word_length
configuration parameter.
.TP
.B WORD_NORMALIZE_CAPITAL
the word contained capital letters and has been converted 
    to lowercase. This bit is only set
    if the
.I wordlist_lowercase
configuration parameter
    is true.
.TP
.B WORD_NORMALIZE_NUMBER
the word contains digits and the configuration 
    parameter
.I wordlist_allow_numbers
is set to false.
.TP
.B WORD_NORMALIZE_CONTROL
the word contains control characters.
.TP
.B WORD_NORMALIZE_BAD
the word is listed in the file pointed by 
    the
.I wordlist_bad_word_list
configuration parameter.
.TP
.B WORD_NORMALIZE_NULL
the word is a zero length string.
.TP
.B WORD_NORMALIZE_PUNCTUATION
at least one character listed in 
    the
.I wordlist_valid_punctuation
attribute was removed
    from the word.
.TP
.B WORD_NORMALIZE_NOALPHA
the word does not contain any alphanumerical character.
.PP
.TP
.B   static String NormalizeStatus(int flags)
Returns a string explaining the return flags of the Normalize
method.
.PP

.SH AUTHORS
Loic Dachary loic@gnu.org

The Ht://Dig group http://dev.htdig.org/


.SH SEE ALSO
htdb_dump(1), htdb_stat(1), htdb_load(1), mifluzdump(1), mifluzload(1), mifluzsearch(1), mifluzdict(1), WordContext(3), WordList(3), WordDict(3), WordListOne(3), WordKey(3), WordKeyInfo(3), WordDBInfo(3), WordRecordInfo(3), WordRecord(3), WordReference(3), WordCursor(3), WordCursorOne(3), WordMonitor(3), Configuration(3), mifluz(3)



gnu_dataset/mifluz-0.26.0/mifluz/WordType.cc
//
// WordType.cc
//
// WordType:  Wrap some attributes to make is...() type
//            functions and other common functions without having to manage
//            the attributes or the exact attribute combination semantics.
//	      Configuration parameter used:
//            valid_punctuation,extra_word_characters,minimum_word_length,
//            maximum_word_length,allow_numbers,bad_word_list
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1999, 2000, 2001 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU General Public License version 2 or later 
// <http://www.gnu.org/copyleft/gpl.html>
//
// $Id: WordType.cc,v 1.11 2001/06/29 14:14:08 loic Exp $
//

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <ctype.h>
#include <stdio.h>
#include <locale.h>

#include "WordType.h"

WordType::WordType(const Configuration &config)
{
  const String valid_punct = config["wordlist_valid_punctuation"];
  const String extra_word_chars = config["wordlist_extra_word_characters"];

  String locale = config.Find("wordlist_locale");
  if(locale.empty())
    locale = "C";
  if(setlocale(LC_ALL, (char*)locale) == 0) {
    fprintf(stderr, "WordType::WordType: cannot set locale: ");
    perror("");
  }

  minimum_length = config.Value("wordlist_minimum_word_length", 3);
  maximum_length = config.Value("wordlist_maximum_word_length", 25);
  allow_numbers = config.Boolean("wordlist_allow_numbers", 0);
  lowercase = config.Boolean("wordlist_lowercase", 1);
  truncate = config.Boolean("wordlist_truncate", 1);

  extra_word_characters = extra_word_chars;
  valid_punctuation = valid_punct;

  chrtypes[0] = 0;
  for (int i = 1; i < 256; i++)
  {
    chrtypes[i] = 0;
    if (isalpha(i))
	chrtypes[i] |= WORD_TYPE_ALPHA;
    if (isdigit(i))
	chrtypes[i] |= WORD_TYPE_DIGIT;
    if (iscntrl(i))
	chrtypes[i] |= WORD_TYPE_CONTROL;
    if (strchr(extra_word_chars, i))
	chrtypes[i] |= WORD_TYPE_EXTRA;
    if (strchr(valid_punct, i))
	chrtypes[i] |= WORD_TYPE_VALIDPUNCT;
  }

  {
    const String filename = config["wordlist_bad_word_list"];
    FILE	*fl = fopen(filename, "r");
    char	buffer[1000];
    char	*word;
    String      new_word;

    // Read in the badwords file (it's just a text file)
    while (fl && fgets(buffer, sizeof(buffer), fl))
      {
	word = strtok(buffer, "\r\n \t");
	if (word && *word)
	  {
	    int flags;
	    new_word = word;
	    if((flags = Normalize(new_word)) & WORD_NORMALIZE_NOTOK) {
	      fprintf(stderr, "WordType::WordType: reading bad words from %s found %s, ignored because %s\n", (const char*)filename, word, (char*)NormalizeStatus(flags & WORD_NORMALIZE_NOTOK));
	    } else {
	      badwords.Add(new_word, 0);
	    }
	  }
    }

    if (fl)
	fclose(fl);
  }
}

int
WordType::Normalize(String& word) const
{
  int status = 0;

  //
  // Reject empty strings, always
  //
  if(word.empty())
    return status | WORD_NORMALIZE_NULL | WORD_NORMALIZE_NOTOK;

  //
  // Always convert to lowercase
  //
  if(lowercase && word.lowercase())
    status |= WORD_NORMALIZE_CAPITAL;

  //
  // Remove punctuation characters according to configuration
  //
  if(StripPunctuation(word))
    status |= WORD_NORMALIZE_PUNCTUATION;

  //
  // Truncate words too long 
  //
  if(word.length() > maximum_length) {
    status |= WORD_NORMALIZE_TOOLONG;
    if(truncate)
      word.chop(word.length() - maximum_length);
    else
      return status | WORD_NORMALIZE_NOTOK;
  }

  //
  // Reject words too short according to configuration
  //
  if(word.length() < minimum_length)
    return status | WORD_NORMALIZE_TOOSHORT | WORD_NORMALIZE_NOTOK;

  //
  // Reject if contains control characters
  //
  int alpha = 0;
  for(const unsigned char *p = (const unsigned char*)(const char*)(char *)word; *p; p++) {
    if(IsStrictChar(*p) || (allow_numbers && IsDigit(*p))) {
      alpha = 1;
    } else if(IsControl(*p)) {
      return status | WORD_NORMALIZE_CONTROL | WORD_NORMALIZE_NOTOK;
    } else if(IsDigit(*p)) {
      return status | WORD_NORMALIZE_NUMBER | WORD_NORMALIZE_NOTOK;
    }
  }

  //
  // Reject if contains no alpha characters (according to configuration)
  //
  if(!alpha) return status | WORD_NORMALIZE_NOALPHA | WORD_NORMALIZE_NOTOK;

  //
  // Reject if listed in config[bad_word_list]
  //
  if(badwords.Exists(word))
    return status | WORD_NORMALIZE_BAD | WORD_NORMALIZE_NOTOK;

  //
  // Accept and report the transformations that occured
  //
  return status | WORD_NORMALIZE_OK;
}

//
// Convert the integer status into a readable string
//
String
WordType::NormalizeStatus(int flags)
{
  String tmp;

  if(flags & WORD_NORMALIZE_TOOLONG) tmp << "TOOLONG ";
  if(flags & WORD_NORMALIZE_TOOSHORT) tmp << "TOOSHORT ";
  if(flags & WORD_NORMALIZE_CAPITAL) tmp << "CAPITAL ";
  if(flags & WORD_NORMALIZE_NUMBER) tmp << "NUMBER ";
  if(flags & WORD_NORMALIZE_CONTROL) tmp << "CONTROL ";
  if(flags & WORD_NORMALIZE_BAD) tmp << "BAD ";
  if(flags & WORD_NORMALIZE_NULL) tmp << "NULL ";
  if(flags & WORD_NORMALIZE_PUNCTUATION) tmp << "PUNCTUATION ";
  if(flags & WORD_NORMALIZE_NOALPHA) tmp << "NOALPHA ";

  if(tmp.empty()) tmp << "GOOD";

  return tmp;
}
